// Copyright 2025 OpenADP Authors.  This work is licensed under the Apache 2.0 license.
syntax = "proto3";

package openadp;

service OpenADP {
    rpc RegisterSecret(RegisterSecretRequest) returns (RegisterSecretResponse) {}
    rpc RecoverSecret(RecoverSecretRequest) returns (RecoverSecretResponse) {}
    rpc ListBackups(ListBackupsRequest) returns (ListBackupsResponse) {}
}

message RegisterSecretRequest {
    string UID = 1;  // User ID
    string DID = 2;  // Device ID
    string BID = 3;  // Backup ID
    int32 version = 4;
    uint32 x = 5;  // Shamir share X coordinate in range [1..N] if there are N shares.
    bytes y = 6;  // Y coordinate, little-endian
    uint32 max_guesses = 7;  // The maximum guesses for the user's recover pin
    uint64 expiration = 8;  // Seconds since 1970.
}

message RegisterSecretResponse {
}

message RecoverSecretRequest {
    string UID = 1;  // User ID
    string DID = 2;  // Device ID
    string BID = 3;  // Backup ID
    bytes B = 4;  // Compressed point. See https://datatracker.ietf.org/doc/html/rfc8032
}

message RecoverSecretResponse {
    uint32 version = 1;
    uint32 x = 2;
    bytes siB = 3;  // y*B, a compressed elliptic curve point.
    uint32 num_guesses = 4;  // How many attempts, including good guesses, that have ever been made.
    uint32 max_guesses = 5;  // Max guesses before we refuse to allow more.
    uint64 expiration = 6;  // Date after which we may delete this share.
}

message ListBackupsRequest {
    repeated Backup backups = 1;
}

message Backup {
    DID = 1;
    BID = 2;
    version = 3;
    num_guesses = 4;
    max_guesses = 5;
    expiration = 6;
}

message ListBackupsResponse {
}
